clippy_lints/src/transmuting_null.rs

```diff
@@ -0,0 +1,96 @@
+use crate::consts::{constant_context, Constant};
+use crate::utils::{match_qpath, span_lint};
+use if_chain::if_chain;
+use rustc::hir::{Expr, ExprKind};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use rustc::{declare_tool_lint, lint_array};
+use syntax::ast::LitKind;
+
+declare_clippy_lint! {
+    pub TRANSMUTING_NULL,
+    correctness,
+    "transmutes from a null pointer to a reference, which is Undefined Behavior"
+}
+
+#[derive(Copy, Clone)]
+pub struct Pass;
+
+impl LintPass for Pass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(TRANSMUTING_NULL,)
+    }
+
+    fn name(&self) -> &'static str {
+        "TransmutingNull"
+    }
+}
+
+const LINT_MSG: &str = "transmuting a known null pointer into a reference.";
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        if_chain! {
```

Comment:
I think this is going to need an `in_external_macro` check as well

Reply:
Why? (I'm not familiar with what `in_external_macro` entails yet :))

Reply:
If you write a macro with `transmute(ZPTR)` in crate_1 and import that macro in crate_2, then it will lint the expanded code. But since the macro is from another crate, where you can't modify the code, you cannot do anything about it.

It should only bail out on _external_ macros though, since you should never write something like `transmute(ZPTR)`, including macros.

Good catch @phansch 👍

Reply:
Niiiiice. I see now :D

Reply:
Added a check based on how the `else_if_without_else.rs` lint does it!